File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/presentation/ui/base/activity/DrawerAdapter.java
package com.lattis.ellipse.presentation.ui.base.activity;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.jakewharton.rxbinding2.view.RxView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.lattis.ellipse.R;
import io.reactivex.subjects.PublishSubject;


public class DrawerAdapter extends RecyclerView.Adapter<DrawerAdapter.ViewHolder> {

    private List<DrawerMenu> drawerMenus;

    DrawerAdapter(List<DrawerMenu> drawerMenus){
        this.drawerMenus=drawerMenus;
    }

    private PublishSubject<View> viewClickSubject = PublishSubject.create();

    private int selectedPosition = 0;


    public void setDrawerMenuList(List<DrawerMenu> drawerMenus){
        this.drawerMenus=drawerMenus;
    }

    public PublishSubject<View> getViewClickSubject() {
        return viewClickSubject;
    }

    public int getSelectedPosition() {
        return this.selectedPosition;
    }

    public void setSelectedPosition(int selectedPosition) {
        this.selectedPosition = selectedPosition;
    }

    @Override
    public int getItemViewType(int position) {
        return 0;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        int viewLayoutId = R.layout.layout_drawer_menu_item_1;

        View view = LayoutInflater.from(parent.getContext()).inflate(viewLayoutId, parent, false);

        ViewHolder viewHolder = new ViewHolder(view);

        if (viewType != 2) {
            RxView.clicks(view)
                    .takeUntil(RxView.detaches(parent))
                    .map(aVoid -> view)
                    .subscribe(viewClickSubject);
        }

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Context context = holder.itemView.getContext();

        DrawerMenu entry = drawerMenus.get(position);

        int titleId = entry.getTitleId();

        if (titleId > 0) {
            holder.titleView.setText(titleId);
        }

        int imageId = entry.getImageId();

        if(imageId>0){
            holder.imageView.setImageDrawable(context.getResources().getDrawable(imageId));
        }

    }

    @Override
    public int getItemCount() {
        return drawerMenus.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        @Nullable
        @BindView(R.id.title)
        TextView titleView;

        @Nullable
        @BindView(R.id.image)
        ImageView imageView;


        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

    }

}

File: velolabs_repos/lattis_android_v2/app/src/main/java/com/lattis/lattis/presentation/base/activity/DrawerAdapter.kt
package com.lattis.lattis.presentation.base.activity

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.jakewharton.rxbinding4.view.clicks
import com.jakewharton.rxbinding4.view.detaches
import com.lattis.domain.models.Reservation
import io.lattis.lattis.R
import io.reactivex.rxjava3.subjects.PublishSubject

class DrawerAdapter internal constructor(
    private var drawerMenus: List<DrawerMenu>
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {


    val viewClickSubject = PublishSubject.create<View>()
    var selectedPosition = 0
    var reservationCount = 0

    companion object{
        val HEADER =0
        val ITEM =1
        val DIVIDER =2
        val RESERVATION =3
    }


    fun setReservationsNumber(reservationCount:Int){
        this.reservationCount = reservationCount
    }

    fun setDrawerMenuList(drawerMenus: List<DrawerMenu>) {
        this.drawerMenus = drawerMenus
    }

    override fun getItemViewType(position: Int): Int {
        val entry = drawerMenus[position]
        return entry.type
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): RecyclerView.ViewHolder {

        if(viewType == ITEM) {
            val viewLayoutId: Int = R.layout.layout_drawer_menu_item_1
            val view =
                LayoutInflater.from(parent.context).inflate(viewLayoutId, parent, false)
            val viewHolder =
                ViewHolderForItem(view)
            if (viewType != 2) {
                view.clicks()
                    .takeUntil(parent.detaches())
                    .map { aVoid: Any? -> view }
                    .subscribe(viewClickSubject)
            }
            return viewHolder
        }else if(viewType == HEADER){
            val viewLayoutId: Int = R.layout.layout_drawer_menu_section_1
            val view =
                LayoutInflater.from(parent.context).inflate(viewLayoutId, parent, false)
            val viewHolder =
                ViewHolderForSection(view)

            return viewHolder
        }else if(viewType == DIVIDER){
            val viewLayoutId: Int = R.layout.layout_drawer_menu_divider_1
            val view =
                LayoutInflater.from(parent.context).inflate(viewLayoutId, parent, false)
            val viewHolder =
                ViewHolderForDivider(view)

            return viewHolder
        }else if (viewType == RESERVATION){
            val viewLayoutId: Int = R.layout.layout_drawer_menu_reservation
            val view =
                LayoutInflater.from(parent.context).inflate(viewLayoutId, parent, false)
            val viewHolder =
                ViewHolderForReservation(view)
            if (viewType != 2) {
                view.clicks()
                    .takeUntil(parent.detaches())
                    .map { aVoid: Any? -> view }
                    .subscribe(viewClickSubject)
            }
            return viewHolder
        }else { // its required to put in else so writing other wise not required
            val viewLayoutId: Int = R.layout.layout_drawer_menu_item_1
            val view =
                LayoutInflater.from(parent.context).inflate(viewLayoutId, parent, false)
            val viewHolder =
                ViewHolderForItem(view)
            if (viewType != 2) {
                view.clicks()
                    .takeUntil(parent.detaches())
                    .map { aVoid: Any? -> view }
                    .subscribe(viewClickSubject)
            }
            return viewHolder
        }
    }

    override fun onBindViewHolder(
        holder: RecyclerView.ViewHolder,
        position: Int
    ) {


        val context = holder.itemView.context
        val entry = drawerMenus[position]

        if(entry.type == ITEM) {

            val titleId = entry.titleId
            if (titleId > 0) {
                (holder as ViewHolderForItem).titleView.setText(titleId)
            }

            var itemId = entry.itemId
            if(itemId!=null && itemId == R.id.menu_reservation && reservationCount>0){
                (holder as ViewHolderForItem).reservationCountView.visibility = View.VISIBLE
                (holder as ViewHolderForItem).reservationCountView.setText(""+reservationCount)
            }else{
                (holder as ViewHolderForItem).reservationCountView.visibility = View.GONE
            }
            val imageId = entry.imageId!!
            if (imageId > 0) {
                (holder as ViewHolderForItem).imageView.setImageDrawable(context.resources.getDrawable(imageId))
            }
        }else if(entry.type == HEADER){
            val titleId = entry.titleId
            if (titleId > 0) {
                (holder as ViewHolderForSection).titleView.setText(titleId)
            }
        }else if(entry.type == DIVIDER){

        }else if(entry.type == RESERVATION
//            && reservationInCurrentStatus!=null &&
//            reservationInCurrentStatus?.reservation_start!=null){
//            val reservationHeading = context.getString(R.string.reservation)+" - "+getCurrentDateFromString(reservationInCurrentStatus?.reservation_start!!)
//            (holder as ViewHolderForReservation).reservationText.text = reservationHeading

            ){
                (holder as ViewHolderForReservation).reservationText.text = context.getString(R.string.reservations)
        }

    }


    override fun getItemCount(): Int {
        return drawerMenus.size
    }

    inner class ViewHolderForItem(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var titleView: TextView = itemView.findViewById(R.id.title)
        var imageView: ImageView = itemView.findViewById(R.id.image)
        var reservationCountView: TextView = itemView.findViewById(R.id.cv_reservation_count)
    }


    inner class ViewHolderForSection(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var titleView: TextView = itemView.findViewById(R.id.title)
    }

    inner class ViewHolderForDivider(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var divider: View = itemView.findViewById(R.id.drawer_divider)
    }

    inner class ViewHolderForReservation(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var reservationText: TextView = itemView.findViewById(R.id.ct_drawer_reservation)
    }



}